ls1intum/Artemis
# Enhancement/scheduling profile

### Checklist

### Motivation and Context
Some tasks within Artemis are scheduled once a day (like cleaning up the build plans) and other tasks are scheduled on a given date (like locking the repositories of programming exercises).
When using a multi-instance deployment, these tasks should only be scheduled once. https://github.com/ls1intum/Artemis/pull/1644 adds a way to schedule quiz exercises dynamically on different nodes and this PR adds a profile `scheduling` that must be enabled to active scheduling for the other tasks.

**Note:** This does not affect the quiz scheduling as this is part of https://github.com/ls1intum/Artemis/pull/1644

### Description
I've added a profile `scheduling` that enables scheduling. Also, to communicate changes made on a programming or text exercise on one instance to the instance that performs the scheduling using a Hazelcast topic.
The instance with the active `scheduling` subscribes to the topics and the instances without the profile `scheduling` publish updates to the subscribed instance.
This subscribed instance will then handle the scheduling, e.g. for new programing exercises.

### Steps for Testing
- Review the code
- If possible, start two instances of Artemis (make sure they are connected within a Hazelcast cluster) and create/update a programming exercise on both the instance with the `scheduling` profile and the instance without the `scheduling` profile.
- Only the instance with the `scheduling` should process the tasks
- If the exercise was created on an instance without the `scheduling` profile enabled, a message should be printed to the log that looks similar to `DistributedInstanceMessageSendService : Sending schedule for programming exercise 47 to broker.` (exerciseId and exercise type may be different)
- The instance with the `scheduling` profile enabled should print a message similar to `InstanceMessageReceiveService : Received schedule update for programming exercise 47` in the log
